sycl/doc/SpecializationConstants.md

```diff
@@ -0,0 +1,307 @@
+    cgh.single_task<MyKernel>(
+        p.get_kernel<MyKernel>(),
+        [=]() {
+          acc[0] = i32.get();
+        });
+  });
+...
+```
+
+## Design
+
+This section describes the basic design used to support spec constants of
+primitive numeric types. POD types support is described further in the document.
+
+#### Compiler
+
+Key `spec_constant::get()` function implementation for the device code:
+
+```cpp
+template <typename T, typename ID = T> class spec_constant {
+...
+public:
+  T get() const { // explicit access.
+#ifdef __SYCL_DEVICE_ONLY__
+    const char *TName = __builtin_unique_stable_name(ID);
+    return __sycl_getSpecConstantValue<T>(TName);
+#else
+    return Val;
+#endif // __SYCL_DEVICE_ONLY__
+  }
+```
+
+here `__builtin_unique_stable_name` is a compiler built-in used to translate
+types to unique strings. `__sycl_getSpecConstantValue<T>` is an "intrinsic"
+recognized by a special LLVM pass later.
+
+Compilation and subsequent linkage of device code results in a number of
+`__sycl_getSpecConstantValue` calls whose arguments are symbolic spec constant
+IDs. Before generating the a device binary, each linked device code LLVMIR
+module undergoes processing by the sycl-post-link tool which can run LLVMIR
+passes before passing the module onto the llvm-spirv translator.
+
+There is a `SpecConstants` LLVMIR pass which
+- assigns numeric IDs to the spec constants
+- brings IR to the form expected by the llvm-spirv translator
+- collects and provides \<Symbolic ID\> =\> \<numeric ID\> spec constant information
+  to the sycl-post-link tool
+Particularly, it translates intrinsic calls to the
+`T __sycl_getSpecConstantValue*(const char *symbolic_id)` intrinsic into
+calls to `T __spirv_SpecConstant(int ID, T default_val)` intrinsic known to
+the llvm-spirv translator. Where `ID` is the numeric ID of the corresponding
+spec constant, `default_val` is its default value which will be used if the
+constant is not set at the runtime.
+
+After this pass the sycl-post-link tool will output the
+\<Symbolic ID\> =\> \<numeric ID\> spec constant mapping into a file for later
+attaching this info to the device binary image via the offload wrapper tool as
+a property set:
+
+```cpp
+struct pi_device_binary_struct {
+...
+  // Array of preperty sets; e.g. specialization constants symbol-int ID map is
+  // propagated to runtime with this mechanism.
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+```
+
+SYCL runtime can then load and access info about particular spec constant using
+its name as a key into the appropriate property set (named "SYCL/specialization
+constants").
+
+##### Ahead of time compilation
+
+With AOT everything is simplified - the `SpecConstants` pass simply replaces
+the `__sycl_getSpecConstantValue` calls with constants - default values of
+the spec constant's type. No maps are generated, and SYCL program can't change
+the value of a spec constant.
```

Comment:
I think it makes sense to mention what happens if we try to change value of the spec constant in AOT mode. I guess an exception is thrown in this case, right?

Reply:
Yes. But this is spec, I don't think we need to duplicate it here.